### PR TITLE
#21199 Prevent pull/push from constructor by throwing IllegalStateException

### LINKED ISSUE #21199: Pulling in GraphStageLogic constructor should not NPE, but fail nicely
Given:

```scala
/**
 * Copyright (C) 2014-2016 Lightbend Inc. <http://www.lightbend.com>
 */
package akka.stream.scaladsl

import java.util.concurrent.ThreadLocalRandom.{ current ⇒ random }

import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler }
import akka.stream._
import akka.stream.testkit._
import scala.concurrent.duration._
import scala.concurrent.Await

class FlowMapSpec extends StreamSpec {

  val settings = ActorMaterializerSettings(system)
    .withInputBuffer(initialSize = 2, maxSize = 16)

  implicit val materializer = ActorMaterializer(settings)

  "A Map" must {

    final class PrintlnSink(prefix: String) extends GraphStage[SinkShape[Any]] {
      final val in = Inlet[Any]("PrintlnSink.in")
      override val shape = SinkShape(in)

      override def createLogic(inheritedAttributes: Attributes) =
        new GraphStageLogic(shape) {
          // a new GraphStageLogic will be created each time we materialize the sink
          // so all state that we want to keep for the stage should be withing the logic.
          // 
          // here we maintain a counter of 
          private var count = 0L

          setHandler(in, new InHandler {
            /**
             * Called when the input port has a new element available.
             * The actual element can be retrieved via the [[GraphStageLogic.grab()]] method.
             */
            override def onPush(): Unit = {
              val element = grab(in)
              count += 1 // the GraphStage maintains the Actor-like single-threaded illusion
              println(s"[$prefix:$count] $element")
              pull(in)
            }
          })

            pull(in) // BOOM! Should have been in preStart
        }
    }

    "map" in {
      val it = Source.repeat("Hello")
        .take(3)
        .map({ i ⇒ println("i = " + i); i })
        .alsoTo(new PrintlnSink("example"))
        .runWith(Sink.ignore)

      Await.ready(it, 3.seconds)
    }

  }

}

```

It fails with a NPE:

```
--- Stream actors debug dump ---
Stream is completed. No debug information is available
[info] - must map *** FAILED *** (2 seconds, 110 milliseconds)
[info]   java.lang.NullPointerException:
[info]   at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:344)
[info]   at akka.stream.scaladsl.FlowMapSpec$$anonfun$1$PrintlnSink$1$$anon$1.<init>(FlowMapSpec.scala:50)
[info]   at akka.stream.scaladsl.FlowMapSpec$$anonfun$1$PrintlnSink$1.createLogic(FlowMapSpec.scala:28)
[info]   at akka.stream.stage.GraphStage.createLogicAndMaterializedValue(GraphStage.scala:43)
[info]   at akka.stream.impl.fusing.GraphInterpreter$GraphAssembly.materialize(GraphInterpreter.scala:192)
[info]   at akka.stream.impl.ActorMaterializerImpl$$anon$2.matGraph(ActorMaterializerImpl.scala:214)
[info]   at akka.stream.impl.ActorMaterializerImpl$$anon$2.materializeAtomic(ActorMaterializerImpl.scala:201)
[info]   at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:919)
[info]   at akka.stream.impl.MaterializerSession$$anonfun$materializeModule$1.apply(StreamLayout.scala:915)
[info]   at scala.collection.immutable.Set$Set2.foreach(Set.scala:128)
[info]   at akka.stream.impl.MaterializerSession.materializeModule(StreamLayout.scala:915)
[info]   at akka.stream.impl.MaterializerSession.materialize(StreamLayout.scala:882)
[info]   at akka.stream.impl.ActorMaterializerImpl.materialize(ActorMaterializerImpl.scala:240)
[info]   at akka.stream.impl.ActorMaterializerImpl.materialize(ActorMaterializerImpl.scala:138)
[info]   at akka.stream.scaladsl.RunnableGraph.run(Flow.scala:351)
[info]   at akka.stream.scaladsl.Source.runWith(Source.scala:81)
[info]   at akka.stream.scaladsl.FlowMapSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(FlowMapSpec.scala:59)
[info]   at akka.stream.scaladsl.FlowMapSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(FlowMapSpec.scala:54)
[info]   at akka.stream.scaladsl.FlowMapSpec$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(FlowMapSpec.scala:54)
[info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
```

It should instead fail with a nicer error message.

### PR DESCRIPTION
**Pull request by 2m**
Fixes #21199

### CODE REVIEW

**Comment by drewhk** [comment]
Easy fix then, thanks! :)

**Comment by drewhk** [comment]
LGTM

**Comment by 2m** [comment]
Yes. At first I was thinking how not to pollute hot path with additional checks. But then noticed this. :)

**Comment by johanandren** [comment]
LGTM!

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 1]
```diff
@@ -343,8 +343,8 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
    */
   final protected def pull[T](in: Inlet[T]): Unit = {
     val connection = conn(in)
-    val portState = connection.portState
     val it = interpreter
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: oh, simple fix. nice.

**Comment by ktoso** [comment]
Great, LGTM!

**Comment by patriknw** [comment]
LGTM

**Pull request merged**